### PR TITLE
Add "gate" device class for DEVICE_CLASS_GATE

### PR DESCRIPTION
**Pull request by bdraco**
## Proposed change

Gates are found outside of a structure and are typically part of a fence.

When opening or closing a garage door, it was impossible to tell
if you had hit the button or not even though the underlying state
was reported as "opening". This lead to confusion and multiple
clicks to open a garage door which can cause the door to stop
opening and result in frustration.

Add icons for gate and garage opening and closing states.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

https://github.com/home-assistant/core/pull/33074
https://github.com/home-assistant/developers.home-assistant/pull/444

### CODE REVIEW

**Comment by bdraco** [comment]
Looks like its preferred to do it the way zwave does it in `case "zwave"` in `domain_icon.ts`

**Comment by bdraco** [comment]
Screen Shot 2020-03-20 at 9 05 26 PM
Screen Shot 2020-03-20 at 9 04 21 PM
Screen Shot 2020-03-20 at 9 04 19 PM
Screen Shot 2020-03-20 at 9 04 16 PM
Screen Shot 2020-03-20 at 9 04 11 PM

**Comment by PeteBa** [comment]
Would it be more intuitive to use icons like below to control the gate ? Whether to press the up or down arrow to open a gate may not be immediately obvious.

image

`mdi-expand-horizontal` and `mdi-collapse-horizontal`

**Comment by bdraco** [comment]
> Would it be more intuitive to use icons like below to control the gate ? Whether to press the up or down arrow to open a gate may not be immediately obvious.
> 
> image
> 
> `mdi-expand-horizontal` and `mdi-collapse-horizontal`

While that would be a nice improvement it would increase the scope of this PR quite a bit as `src/components/ha-cover-controls.js` doesn't contemplate per device class controls.

File: src/common/entity/cover_icon.ts [hunk 1]
```diff
@@ -4,17 +4,58 @@ import { domainIcon } from "./domain_icon";
 
 export const coverIcon = (state: HassEntity): string => {
   const open = state.state !== "closed";
+
   switch (state.attributes.device_class) {
     case "garage":
-      return open ? "hass:garage-open" : "hass:garage";
+      switch (state.state) {
+        case "opening":
+          return "hass:arrow-up-bold-box";
```
Comment (bramkragten) [inline src/common/entity/cover_icon.ts]: I think we shouldn't go for the `bold` variant to align with the style of the buttons

Comment (bdraco) [reply src/common/entity/cover_icon.ts]: Good call. Will adjust

Comment (bdraco) [reply src/common/entity/cover_icon.ts]: That took longer to test than I expected. It looks like I have to restart scripts/develop for it to see new icons

Comment (bramkragten) [reply src/common/entity/cover_icon.ts]: Yes, it generates the `hass` icon set when you start `develop`

**Review by balloob — APPROVED** [review]

**Pull request merged**